### PR TITLE
Many known geometry transforms fixed.

### LINKED ISSUE #45: Low-level handling of degenerate polygons
`Projection._project_polygon()` is rather unhelpfully converting degenerate polygons (such as those produced by `matplotlib.contourf`) into global extent.
e.g.
```Polygon([(20, 20), (20, 20), (20, 20), (20, 20)])```
```Polygon([(20, 20), (20, 20), (20, 20.1), (20, 20)])```

Whilst it's easy to filter these out in the MPL integration layer, ideally the low level re-projection code should be able to deal with them.

### PR DESCRIPTION
**Pull request by pelson**
Closes #45, #259 (both cases within) and #541.

Does not close: #77 (adding a set_global shows that the transform isn't perfect), #325

Need to rationalise (i.e. reduce to single cases): #318.

A little further work will render #492 fixable (already it is better, as it stops the globe being filled incorrectly, but Antarctica is being filled).

### CODE REVIEW

**Comment by pelson** [comment]
As a follow-up PR, I intend to move most of this code out of crs.py.

File: lib/cartopy/crs.py [hunk 1]
```diff
@@ -338,43 +339,98 @@ def boundary_distance(xy):
         # Record the positions of all the segment ends
         for i, line_string in enumerate(line_strings):
             first_dist = boundary_distance(line_string.coords[0])
-            thing = _Thing(first_dist, False,
-                           (i, 'first', line_string.coords[0]))
+            thing = _BoundaryPoint(first_dist, False,
```
Comment (rhattersley) [inline lib/cartopy/crs.py]: :+1: Oh boy, has that been a long time coming!

Comment (rhattersley) [inline lib/cartopy/crs.py]: Although I guess it might be nice to rename the `thing` variables as well.

Comment (pelson) [inline lib/cartopy/crs.py]: Refactor needed - I'm just going to lift the whole lot out of crs.py. And yes, there will be no mention of "thing" anywhere :wink: .

File: lib/cartopy/crs.py [hunk 2]
```diff
@@ -338,43 +339,98 @@ def boundary_distance(xy):
...
             edge_things.append(thing)
             last_dist = boundary_distance(line_string.coords[-1])
-            thing = _Thing(last_dist, False,
-                           (i, 'last', line_string.coords[-1]))
+            if last_dist == first_dist:
+                # We have a closed line-string. Just offset the last_dist by a
+                # tiny fraction.
+                last_dist += 1e-8  # TODO: Use epsilon.
+                thing = _BoundaryPoint(last_dist, False,
+                                       (i, 'self-closing',
```
Comment (pelson) [inline lib/cartopy/crs.py]: It seems this concept isn't needed now I've dealt with co-incident _BoundaryPoint sections.

**Comment by pelson** [comment]
I'm tempted to bank this, and look at any remaining issues separately. As this is a slightly more involved change would you be able to take a look at some point @rhattersley? If not in the short term, I'll continue working on this PR until perhaps Friday afternoon/Monday.

Cheers,

**Comment by esc24** [comment]
Just looking a the visual result, this clearly improves things, but it's interesting that the top and bottom of the IGH boundary are flattened.

**Comment by pelson** [comment]
> Just looking a the visual result, this clearly improves things, but it's interesting that the top and bottom of the IGH boundary are flattened.

Yeah, agreed. Curious, but not outright worrying.

**Comment by pelson** [comment]
Having spoken to @rhattersley (who is the only person who with the kind of knowledge of the algorithm needed to review this algorithmically) I think this should be merged. If there are any general (i.e. non-algorithmic) improvements that can be suggested based on this PR, I'd be happy to address them in a follow-up PR.

**Pull request merged**